File: roulette.cs
// roulette.cs - Симулятор европейской рулетки на C#
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;

class Roulette
{
    private static readonly int[] Numbers = Enumerable.Range(0, 37).ToArray();
    private static readonly HashSet<int> RedNumbers = new HashSet<int> {1,3,5,7,9,12,14,16,18,19,21,23,25,27,30,32,34,36};
    private static readonly HashSet<int> BlackNumbers = new HashSet<int> {2,4,6,8,10,11,13,15,17,20,22,24,26,28,29,31,33,35};

    private int balance;
    private List<Dictionary<string, object>> history;
    private int totalSpins;
    private int totalWins;
    private List<Bet> pendingBets;
    private Random rand;

    private class Bet
    {
        public string Type { get; set; }
        public int Amount { get; set; }
        public int Value { get; set; }
    }

    public Roulette(int initialBalance = 1000)
    {
        balance = initialBalance;
        history = new List<Dictionary<string, object>>();
        totalSpins = 0;
        totalWins = 0;
        pendingBets = new List<Bet>();
        rand = new Random();
        LoadState();
    }

    public int Spin() => Numbers[rand.Next(Numbers.Length)];

    public string GetColor(int number)
    {
        if (number == 0) return "зелёное";
        if (RedNumbers.Contains(number)) return "красное";
        return "чёрное";
    }

    public void PlaceBet(string type, int amount, int value)
    {
        if (amount > balance) throw new Exception("Недостаточно средств");
        balance -= amount;
        pendingBets.Add(new Bet { Type = type, Amount = amount, Value = value });
    }

    public int ResolveBets(int number)
    {
        int totalWin = 0;
        string color = GetColor(number);
        foreach (var bet in pendingBets)
        {
            int win = CalcWin(bet, number, color);
            totalWin += win;
            balance += win;
            var entry = new Dictionary<string, object>
            {
                ["number"] = number,
                ["color"] = color,
                ["betType"] = bet.Type,
                ["betValue"] = bet.Value,
                ["amount"] = bet.Amount,
                ["win"] = win
            };
            history.Add(entry);
        }
        pendingBets.Clear();
        totalSpins++;
        if (totalWin > 0) totalWins++;
        return totalWin;
    }

    private int CalcWin(Bet bet, int number, string color)
    {
        switch (bet.Type)
        {
            case "STRAIGHT": return bet.Value == number ? bet.Amount * 36 : 0;
            case "COLOR":
                if ((bet.Value == 0 && color == "красное") || (bet.Value == 1 && color == "чёрное"))
                    return bet.Amount * 2;
                return 0;
            case "PARITY":
                if ((bet.Value == 0 && number % 2 == 0 && number != 0) ||
                    (bet.Value == 1 && number % 2 == 1))
                    return bet.Amount * 2;
                return 0;
            case "HIGHLOW":
                if ((bet.Value == 0 && number >= 1 && number <= 18) ||
                    (bet.Value == 1 && number >= 19 && number <= 36))
                    return bet.Amount * 2;
                return 0;
            case "DOZEN":
                if ((bet.Value == 1 && number >= 1 && number <= 12) ||
                    (bet.Value == 2 && number >= 13 && number <= 24) ||
                    (bet.Value == 3 && number >= 25 && number <= 36))
                    return bet.Amount * 3;
                return 0;
            case "COLUMN":
                if (number == 0) return 0;
                int col = (number - 1) % 3 + 1;
                return bet.Value == col ? bet.Amount * 3 : 0;
            default: return 0;
        }
    }

    public string GetStats()
    {
        if (totalSpins == 0) return "Нет спинов.";
        double winRate = (double)totalWins / totalSpins * 100;
        return $"Всего спинов: {totalSpins}, Выигрышных: {totalWins} ({winRate:F1}%), Баланс: {balance}";
    }

    public void SaveState(string filename = "roulette_state.json")
    {
        var data = new Dictionary<string, object>
        {
            ["balance"] = balance,
            ["history"] = history,
            ["totalSpins"] = totalSpins,
            ["totalWins"] = totalWins
        };
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
    }

    public void LoadState(string filename = "roulette_state.json")
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        try
        {
            var data = JsonSerializer.Deserialize<Dictionary<string, object>>(json);
            if (data != null)
            {
                balance = Convert.ToInt32(data["balance"]);
                totalSpins = Convert.ToInt32(data["totalSpins"]);
                totalWins = Convert.ToInt32(data["totalWins"]);
                // history загружать сложнее из-за вложенности, для простоты пропустим
                // В реальном проекте лучше использовать конкретные классы
            }
        }
        catch { }
    }

    public void DisplayHistory(int n = 10)
    {
        if (history.Count == 0)
        {
            Console.WriteLine("История пуста.");
            return;
        }
        Console.WriteLine("\x1b[36mПоследние спины:\x1b[0m");
        int start = Math.Max(0, history.Count - n);
        for (int i = start; i < history.Count; i++)
        {
            var entry = history[i];
            string color = (string)entry["color"];
            string colorCode = color == "красное" ? "\x1b[31m" :
                               color == "чёрное" ? "\x1b[30m" : "\x1b[32m";
            Console.WriteLine($"{entry["number"]} ({colorCode}{color}\x1b[0m) Ставка: {entry["betType"]} {entry["betValue"]} Сумма: {entry["amount"]} -> Выигрыш: {entry["win"]}");
        }
    }

    public int Balance => balance;
    public bool HasPendingBets => pendingBets.Count > 0;
}

class Program
{
    static void Main(string[] args)
    {
        if (args.Length > 0 && args[0] == "--auto")
        {
            var roulette = new Roulette();
            Console.WriteLine("Автоигра: ставка на красное по 10 фишек, 20 спинов");
            for (int i = 0; i < 20; i++)
            {
                if (roulette.Balance < 10) break;
                roulette.PlaceBet("COLOR", 10, 0);
                int number = roulette.Spin();
                int win = roulette.ResolveBets(number);
                Console.WriteLine($"Спин {i+1}: {number} {roulette.GetColor(number)}, выигрыш: {win}, баланс: {roulette.Balance}");
                Thread.Sleep(500);
            }
            roulette.SaveState();
            Console.WriteLine("Автоигра завершена.");
            return;
        }

        // Интерактивный режим
        var r = new Roulette();
        Console.WriteLine("\x1b[33m🎰 Добро пожаловать в Европейскую рулетку!\x1b[0m");
        while (true)
        {
            Console.WriteLine($"\nБаланс: {r.Balance} фишек");
            Console.WriteLine("\nДоступные ставки:");
            Console.WriteLine("1. Прямая (номер)");
            Console.WriteLine("2. Цвет (красное/чёрное)");
            Console.WriteLine("3. Чёт/Нечет");
            Console.WriteLine("4. Большое/Малое");
            Console.WriteLine("5. Дюжина");
            Console.WriteLine("6. Колонка");
            Console.WriteLine("7. Спин");
            Console.WriteLine("8. Статистика");
            Console.WriteLine("9. История");
            Console.WriteLine("10. Сохранить и выйти");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine().Trim();

            if (choice == "10")
            {
                r.SaveState();
                Console.WriteLine("Состояние сохранено. До свидания!");
                break;
            }
            else if (choice == "8")
            {
                Console.WriteLine("\x1b[35m" + r.GetStats() + "\x1b[0m");
                continue;
            }
            else if (choice == "9")
            {
                r.DisplayHistory();
                continue;
            }
            else if (choice == "7")
            {
                if (!r.HasPendingBets)
                {
                    Console.WriteLine("\x1b[31mСначала сделайте ставку!\x1b[0m");
                    continue;
                }
                Console.WriteLine("Крутим колесо...");
                Thread.Sleep(1000);
                int number = r.Spin();
                string color = r.GetColor(number);
                string colorCode = color == "красное" ? "\x1b[31m" :
                                   color == "чёрное" ? "\x1b[30m" : "\x1b[32m";
                Console.WriteLine($"Выпало: {number} ({colorCode}{color}\x1b[0m)");
                int win = r.ResolveBets(number);
                if (win > 0) Console.WriteLine($"\x1b[32mВы выиграли {win} фишек!\x1b[0m");
                else Console.WriteLine("\x1b[31mВы проиграли ставку.\x1b[0m");
                continue;
            }

            // Ставки
            if (!int.TryParse(choice, out int betType) || betType < 1 || betType > 6)
            {
                Console.WriteLine("\x1b[31mНеверный выбор.\x1b[0m");
                continue;
            }
            Console.Write("Сумма ставки: ");
            string amountStr = Console.ReadLine().Trim();
            if (!int.TryParse(amountStr, out int amount) || amount <= 0 || amount > r.Balance)
            {
                Console.WriteLine("\x1b[31mНекорректная сумма.\x1b[0m");
                continue;
            }
            int value = 0;
            bool ok = true;
            switch (betType)
            {
                case 1:
                    Console.Write("Введите номер (0-36): ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || value < 0 || value > 36)
                    { Console.WriteLine("\x1b[31mНомер от 0 до 36.\x1b[0m"); ok = false; }
                    else r.PlaceBet("STRAIGHT", amount, value);
                    break;
                case 2:
                    Console.Write("0 - красное, 1 - чёрное: ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || (value != 0 && value != 1))
                    { Console.WriteLine("\x1b[31mВведите 0 или 1.\x1b[0m"); ok = false; }
                    else r.PlaceBet("COLOR", amount, value);
                    break;
                case 3:
                    Console.Write("0 - чёт, 1 - нечет: ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || (value != 0 && value != 1))
                    { Console.WriteLine("\x1b[31mВведите 0 или 1.\x1b[0m"); ok = false; }
                    else r.PlaceBet("PARITY", amount, value);
                    break;
                case 4:
                    Console.Write("0 - малое (1-18), 1 - большое (19-36): ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || (value != 0 && value != 1))
                    { Console.WriteLine("\x1b[31mВведите 0 или 1.\x1b[0m"); ok = false; }
                    else r.PlaceBet("HIGHLOW", amount, value);
                    break;
                case 5:
                    Console.Write("Дюжина (1,2,3): ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || value < 1 || value > 3)
                    { Console.WriteLine("\x1b[31mВведите 1,2 или 3.\x1b[0m"); ok = false; }
                    else r.PlaceBet("DOZEN", amount, value);
                    break;
                case 6:
                    Console.Write("Колонка (1,2,3): ");
                    if (!int.TryParse(Console.ReadLine().Trim(), out value) || value < 1 || value > 3)
                    { Console.WriteLine("\x1b[31mВведите 1,2 или 3.\x1b[0m"); ok = false; }
                    else r.PlaceBet("COLUMN", amount, value);
                    break;
            }
            if (ok) Console.WriteLine("\x1b[32mСтавка принята.\x1b[0m");
        }
    }
}

File: roulette.rs
// roulette.rs - Симулятор европейской рулетки на Rust
// Для запуска требуется Cargo.toml с зависимостями:
// [dependencies]
// rand = "0.8"
// serde = { version = "1.0", features = ["derive"] }
// serde_json = "1.0"

use std::collections::HashSet;
use std::io::{self, Write, BufRead};
use std::fs;
use rand::Rng;
use serde::{Serialize, Deserialize};

const RED_NUMBERS: [i32; 18] = [1,3,5,7,9,12,14,16,18,19,21,23,25,27,30,32,34,36];
const BLACK_NUMBERS: [i32; 18] = [2,4,6,8,10,11,13,15,17,20,22,24,26,28,29,31,33,35];

lazy_static::lazy_static! {
    static ref RED_SET: HashSet<i32> = RED_NUMBERS.iter().cloned().collect();
    static ref BLACK_SET: HashSet<i32> = BLACK_NUMBERS.iter().cloned().collect();
}

#[derive(Clone, Serialize, Deserialize)]
struct Bet {
    bet_type: String,
    amount: i32,
    value: i32,
}

#[derive(Serialize, Deserialize)]
struct Roulette {
    balance: i32,
    history: Vec<serde_json::Value>,
    total_spins: i32,
    total_wins: i32,
    #[serde(skip)]
    pending_bets: Vec<Bet>,
    #[serde(skip)]
    rng: rand::rngs::ThreadRng,
}

impl Roulette {
    fn new(initial_balance: i32) -> Self {
        let mut r = Self {
            balance: initial_balance,
            history: Vec::new(),
            total_spins: 0,
            total_wins: 0,
            pending_bets: Vec::new(),
            rng: rand::thread_rng(),
        };
        r.load_state();
        r
    }

    fn spin(&mut self) -> i32 {
        self.rng.gen_range(0..=36)
    }

    fn get_color(&self, number: i32) -> String {
        if number == 0 { "зелёное".to_string() }
        else if RED_SET.contains(&number) { "красное".to_string() }
        else { "чёрное".to_string() }
    }

    fn place_bet(&mut self, bet_type: &str, amount: i32, value: i32) -> Result<(), String> {
        if amount > self.balance {
            return Err("Недостаточно средств".to_string());
        }
        self.balance -= amount;
        self.pending_bets.push(Bet {
            bet_type: bet_type.to_string(),
            amount,
            value,
        });
        Ok(())
    }

    fn resolve_bets(&mut self, number: i32) -> i32 {
        let mut total_win = 0;
        let color = self.get_color(number);
        let bets = std::mem::take(&mut self.pending_bets);
        for bet in bets {
            let win = self.calc_win(&bet, number, &color);
            total_win += win;
            self.balance += win;
            let entry = serde_json::json!({
                "number": number,
                "color": color,
                "betType": bet.bet_type,
                "betValue": bet.value,
                "amount": bet.amount,
                "win": win
            });
            self.history.push(entry);
        }
        self.total_spins += 1;
        if total_win > 0 {
            self.total_wins += 1;
        }
        total_win
    }

    fn calc_win(&self, bet: &Bet, number: i32, color: &str) -> i32 {
        match bet.bet_type.as_str() {
            "STRAIGHT" => if bet.value == number { bet.amount * 36 } else { 0 },
            "COLOR" => {
                if (bet.value == 0 && color == "красное") || (bet.value == 1 && color == "чёрное") {
                    bet.amount * 2
                } else { 0 }
            }
            "PARITY" => {
                if (bet.value == 0 && number % 2 == 0 && number != 0) ||
                   (bet.value == 1 && number % 2 == 1) {
                    bet.amount * 2
                } else { 0 }
            }
            "HIGHLOW" => {
                if (bet.value == 0 && number >= 1 && number <= 18) ||
                   (bet.value == 1 && number >= 19 && number <= 36) {
                    bet.amount * 2
                } else { 0 }
            }
            "DOZEN" => {
                if (bet.value == 1 && number >= 1 && number <= 12) ||
                   (bet.value == 2 && number >= 13 && number <= 24) ||
                   (bet.value == 3 && number >= 25 && number <= 36) {
                    bet.amount * 3
                } else { 0 }
            }
            "COLUMN" => {
                if number == 0 { return 0; }
                let col = (number - 1) % 3 + 1;
                if bet.value == col { bet.amount * 3 } else { 0 }
            }
            _ => 0,
        }
    }

    fn get_stats(&self) -> String {
        if self.total_spins == 0 {
            return "Нет спинов.".to_string();
        }
        let win_rate = self.total_wins as f64 / self.total_spins as f64 * 100.0;
        format!("Всего спинов: {}, Выигрышных: {} ({:.1}%), Баланс: {}",
                self.total_spins, self.total_wins, win_rate, self.balance)
    }

    fn save_state(&self, filename: &str) -> Result<(), String> {
        let data = serde_json::json!({
            "balance": self.balance,
            "history": self.history,
            "total_spins": self.total_spins,
            "total_wins": self.total_wins,
        });
        let json = serde_json::to_string_pretty(&data).map_err(|e| e.to_string())?;
        fs::write(filename, json).map_err(|e| e.to_string())
    }

    fn load_state(&mut self) {
        if let Ok(data) = fs::read_to_string("roulette_state.json") {
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(&data) {
                if let Some(b) = json["balance"].as_i64() { self.balance = b as i32; }
                if let Some(s) = json["total_spins"].as_i64() { self.total_spins = s as i32; }
                if let Some(w) = json["total_wins"].as_i64() { self.total_wins = w as i32; }
                if let Some(h) = json["history"].as_array() {
                    self.history = h.clone();
                }
            }
        }
    }

    fn display_history(&self, n: usize) {
        if self.history.is_empty() {
            println!("История пуста.");
            return;
        }
        println!("\x1b[36mПоследние спины:\x1b[0m");
        let start = if self.history.len() > n { self.history.len() - n } else { 0 };
        for i in start..self.history.len() {
            let entry = &self.history[i];
            let num = entry["number"].as_i64().unwrap_or(0);
            let color = entry["color"].as_str().unwrap_or("");
            let color_code = if color == "красное" { "\x1b[31m" } else if color == "чёрное" { "\x1b[30m" } else { "\x1b[32m" };
            println!("{} ({}{}\x1b[0m) Ставка: {} {} Сумма: {} -> Выигрыш: {}",
                     num, color_code, color,
                     entry["betType"].as_str().unwrap_or(""),
                     entry["betValue"].as_i64().unwrap_or(0),
                     entry["amount"].as_i64().unwrap_or(0),
                     entry["win"].as_i64().unwrap_or(0));
        }
    }

    fn has_pending_bets(&self) -> bool { !self.pending_bets.is_empty() }
}

fn read_line() -> String {
    let stdin = io::stdin();
    let mut line = String::new();
    stdin.lock().read_line(&mut line).expect("Ошибка ввода");
    line.trim().to_string()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && args[1] == "--auto" {
        let mut roulette = Roulette::new(1000);
        println!("Автоигра: ставка на красное по 10 фишек, 20 спинов");
        for i in 0..20 {
            if roulette.balance < 10 { break; }
            roulette.place_bet("COLOR", 10, 0).unwrap();
            let number = roulette.spin();
            let win = roulette.resolve_bets(number);
            println!("Спин {}: {} {}, выигрыш: {}, баланс: {}", i+1, number, roulette.get_color(number), win, roulette.balance);
            std::thread::sleep(std::time::Duration::from_millis(500));
        }
        roulette.save_state("roulette_state.json").unwrap();
        println!("Автоигра завершена.");
        return;
    }

    let mut roulette = Roulette::new(1000);
    println!("\x1b[33m🎰 Добро пожаловать в Европейскую рулетку!\x1b[0m");
    loop {
        println!("\nБаланс: {} фишек", roulette.balance);
        println!("\nДоступные ставки:");
        println!("1. Прямая (номер)");
        println!("2. Цвет (красное/чёрное)");
        println!("3. Чёт/Нечет");
        println!("4. Большое/Малое");
        println!("5. Дюжина");
        println!("6. Колонка");
        println!("7. Спин");
        println!("8. Статистика");
        println!("9. История");
        println!("10. Сохранить и выйти");
        print!("Выберите действие: ");
        io::stdout().flush().unwrap();
        let choice = read_line();

        match choice.as_str() {
            "10" => {
                roulette.save_state("roulette_state.json").unwrap();
                println!("Состояние сохранено. До свидания!");
                break;
            }
            "8" => println!("\x1b[35m{}\x1b[0m", roulette.get_stats()),
            "9" => roulette.display_history(10),
            "7" => {
                if !roulette.has_pending_bets() {
                    println!("\x1b[31mСначала сделайте ставку!\x1b[0m");
                    continue;
                }
                println!("Крутим колесо...");
                std::thread::sleep(std::time::Duration::from_secs(1));
                let number = roulette.spin();
                let color = roulette.get_color(number);
                let color_code = if color == "красное" { "\x1b[31m" } else if color == "чёрное" { "\x1b[30m" } else { "\x1b[32m" };
                println!("Выпало: {} ({}{}\x1b[0m)", number, color_code, color);
                let win = roulette.resolve_bets(number);
                if win > 0 {
                    println!("\x1b[32mВы выиграли {} фишек!\x1b[0m", win);
                } else {
                    println!("\x1b[31mВы проиграли ставку.\x1b[0m");
                }
            }
            _ => {
                // Ставки
                let bet_type: i32 = choice.parse().unwrap_or(0);
                if bet_type < 1 || bet_type > 6 {
                    println!("\x1b[31mНеверный выбор.\x1b[0m");
                    continue;
                }
                print!("Сумма ставки: ");
                io::stdout().flush().unwrap();
                let amount_str = read_line();
                let amount: i32 = amount_str.parse().unwrap_or(0);
                if amount <= 0 || amount > roulette.balance {
                    println!("\x1b[31mНекорректная сумма.\x1b[0m");
                    continue;
                }
                let mut value = 0;
                let mut ok = true;
                match bet_type {
                    1 => {
                        print!("Введите номер (0-36): ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value < 0 || value > 36 {
                            println!("\x1b[31mНомер от 0 до 36.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("STRAIGHT", amount, value).unwrap();
                        }
                    }
                    2 => {
                        print!("0 - красное, 1 - чёрное: ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value != 0 && value != 1 {
                            println!("\x1b[31mВведите 0 или 1.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("COLOR", amount, value).unwrap();
                        }
                    }
                    3 => {
                        print!("0 - чёт, 1 - нечет: ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value != 0 && value != 1 {
                            println!("\x1b[31mВведите 0 или 1.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("PARITY", amount, value).unwrap();
                        }
                    }
                    4 => {
                        print!("0 - малое (1-18), 1 - большое (19-36): ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value != 0 && value != 1 {
                            println!("\x1b[31mВведите 0 или 1.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("HIGHLOW", amount, value).unwrap();
                        }
                    }
                    5 => {
                        print!("Дюжина (1,2,3): ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value < 1 || value > 3 {
                            println!("\x1b[31mВведите 1,2 или 3.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("DOZEN", amount, value).unwrap();
                        }
                    }
                    6 => {
                        print!("Колонка (1,2,3): ");
                        io::stdout().flush().unwrap();
                        value = read_line().parse().unwrap_or(-1);
                        if value < 1 || value > 3 {
                            println!("\x1b[31mВведите 1,2 или 3.\x1b[0m");
                            ok = false;
                        } else {
                            roulette.place_bet("COLUMN", amount, value).unwrap();
                        }
                    }
                    _ => {}
                }
                if ok {
                    println!("\x1b[32mСтавка принята.\x1b[0m");
                }
            }
        }
    }
}
